### ISSUE TITLE
M3 Performance

### ISSUE THREAD

**Issue by Narsil** [philipturner/metal-flash-attention#11]
Hi thanks a lot for this really cool library.

I've taken it for a spin on M3, and saw on it that GEMM seemed to perform better on MPS.
Do you have any suggestions as to why ? Any way we could help ?

**Comment by philipturner** [comment]
Apple rewrote the M3's memory architecture from scratch. This may have obviated the SIMD async copy instructions, or not; I don't know. However, MPS looks like it's not reaching optimal performance. Here is a graph from @liuliu for M3 Pro.

![Figure_1](https://github.com/philipturner/metal-flash-attention/assets/71743241/165bcf8f-3c37-4c86-8dbc-5e40136110e9)

I do have an Apple9 device right now (iPhone 15 Pro Max), but the tests can't run on iOS right now. They have a dependency on Python for NumPy benchmarking and Matplotlib. Next summer, I'll need to use something else to visualize matmul kernel bugs and perf charts. Maybe export data to JSON and graph externally, or use SwiftUI + some Metal graphics shaders on-device.

For more context, I raised an issue on the ISA repository before M3 was released. It referred to some strange structure to MPS kernels for the never-released A16 raytracing GPU. https://github.com/dougallj/applegpu/issues/44. I might suggest increasing the accumulator size to something absurd, although that probably won't fully fix the problem.

**Comment by Narsil** [comment]
Here are are mine.

I tried fiddling around with cache sizes but as suspected it didn't yield anything significant.

![F16](https://github.com/philipturner/metal-flash-attention/assets/204321/471325e6-e4d6-46ed-b1e0-eb97bd91099e)
![float32](https://github.com/philipturner/metal-flash-attention/assets/204321/79328834-4a1c-4507-a8e8-a3418748d865)


One surpising thing to me, was that reducing the number of splits to `1` for M and N, gave a much more consistent throughput for the GEMM op.

Happy to provide further test, or even try live debugging if you want to to access indirectly the M3.

Cheers.

For completeness:
```
GPU name: Apple M3 Max
GPU vendor: Apple
GPU core count: 40
GPU clock frequency: 1.398 GHz
GPU bandwidth: 409.6 GB/s
GPU FLOPS: 14.316 TFLOPS
GPU IPS: 7.158 TIPS
GPU system level cache: 48 MB
GPU memory: 128 GB
GPU family: Apple 8
```

**Comment by philipturner** [comment]
For reference, I superimposed the M1 Max graph on Nick's M3 Max graph. The supermassive dips in MPS performance on M1 came from misuse of SIMD async copy instructions. Since the M3 MPS shaders don't use SIMD async copy, the performance is more consistent for uneven matrix sizes. However, MPS is nowhere near optimal, shown by M1 MFA outperforming M3 MPS for M=200-600.

<img width="677" alt="Screenshot 2023-12-09 at 5 29 35 PM" src="https://github.com/philipturner/metal-flash-attention/assets/71743241/9e2564ae-7a85-42d8-8b7d-3ac5c9fad519">

Perhaps the SIMD async copy instructions were removed from M3, and the runtime AIR compiler replaces them with a slow emulated version. My first suggestion would be modifying the kernel to directly load from device memory, which is non-trivial (although the Metal source code is so elegant, it's easier to modify than any other implementation). Usage of SIMD async copies allows use of the 16-bit address range for threadgroup memory pointers and only temporarily materializing 64-bit pointer.

**Comment by philipturner** [comment]
> One surpising thing to me, was that reducing the number of splits to 1 for M and N, gave a much more consistent throughput for the GEMM op.

I think the MPS kernel is structured so a single SIMD group accesses an entire accumulator in isolation. The reason MFA uses splits, is so 4 simd groups can amortize the overhead of the `device` -> `threadgroup` transfer by 2 times. Also, with the M1 architecture, a single SIMD group can't access an accumulator as large as 64 x 64 x FP32 (16 KB). Getting close (48 x 48) would decrease occupancy too much. Perhaps M3 is different.

**Comment by Narsil** [comment]
You're talking about this, right ? https://github.com/philipturner/metal-flash-attention/blob/main/Sources/GEMM.metal#L203-L216

**Comment by philipturner** [comment]
Yeah. The 64-bit pointer is only temporarily materialized, recycling registers allocated for other data during the GEMM part.

**Comment by Narsil** [comment]
If I understand briefly the 64x64 allocator idea, you mean increasing this buffer to 4096:

And updating the MNK_simd sizes here : https://github.com/philipturner/metal-flash-attention/blob/32592c98eff18001d4eec2c7a204e288fa92fa44/Tests/Operations/GEMM.swift#L124-L139 in order to increase the threadgroup memory size ?

I'm not super familiar with the codebase, just trying to attempt simple ideas here. If they've thrown async simd primitives you found in 14.2 out of the windows it'd be a shame, and not sure how I could help at this point.

**Comment by philipturner** [comment]
That 1024 is just to fool the compiler. It doesn't allocate any blocks. You can increase the accumulator size by trying 64x64, 80x80, or 96x96 instead of 48x48.

> I'm not super familiar with the codebase, just trying to attempt simple ideas here. If they've thrown async simd primitives you found in 14.2 out of the windows it'd be a shame, and not sure how I could help at this point.

If they threw it out of the window, it's simply unnecessary. It might require some simplifications to the GPU kernel, which might not be super hard. Just more than changing a single parameter. Requires more than a few hours of time and some familiarity with Metal. Also a new function constant to take the simplified M3 branch on newer devices, while still supporting older ones.

Attention is a more complex kernel, but the advantage over MPS is already enormous there. So the biggest issue is just fixing GEMM, which should be simpler.

**Comment by philipturner** [comment]
Out of curiosity, what is MPS performance for Float16?

**Comment by philipturner** [comment]
Also if it's just 1 split, you could fine-tune (40x40, 56x56, 72x72).

**Comment by philipturner** [comment]
<img width="683" alt="Screenshot 2023-12-09 at 6 58 05 PM" src="https://github.com/philipturner/metal-flash-attention/assets/71743241/4bdfff41-cd7a-4f3e-8c74-63f8dad5b15f">

**Comment by DigitalSolomon** [comment]
> Apple rewrote the M3's memory architecture from scratch. This may have obviated the SIMD async copy instructions, or not; I don't know. However, MPS looks like it's not reaching optimal performance. Here is a graph from @liuliu for M3 Pro.
> 

I wonder if we're in for another memory architecture rewrite with M4.

**Comment by philipturner** [comment]
I doubt it. Also, M3 is not that hard, compared to M1. I just haven’t had any factor motivating me to patch up MFA yet.

**Comment by philipturner** [comment]
https://gist.github.com/philipturner/3bda14e876a635e73745c42f2eb240c8

**Comment by philipturner** [comment]
Closed due to the issue being stale. The entire repository has been overhauled. It targets a very large combinatorial matrix of hardware + problem size + precision + whatever features you add through programming / time dedication.
